### ISSUE TITLE
LAGHOST fails to converge 

### ISSUE THREAD

**Issue by echoi** [sungho91/Laghost#3]
I tried to run Laghost (commit id: `1b499ddb5f3b833c55f1897f4a0a2035333d7e90`)
```SHELL
Author: sungho91 <sungho91@snu.ac.kr>
Date:   Thu Jul 4 18:42:02 2024 -0500

    add_test_dir
```
built with MFEM commit `c0867ca009e9b5fafb5585318648f430b1d6eef0`, but it failed to converge:
```SHELL
simulation starts
PCG: Number of iterations: 100
PCG: No convergence!
```
The full output log can be found in this file: [failed_run.txt](https://github.com/user-attachments/files/16490024/failed_run.txt)

One question is that the max iteration seems set to be 1000 in the used input file, [defaults_SL.cfg.txt](https://github.com/user-attachments/files/16490049/defaults_SL.cfg.txt), but the PCG iteration stops at 100.
 
This issue contradicts the separate [issue #2](https://github.com/sungho91/Laghost/issues/2) that the tried MFEM version works with Laghost.

**Comment by echoi** [comment]
I tried to go back to the baseline but failed. 

I copied `/home/slee29/mfem` and `/home/slee29/Laghost` on `nefertiti` to my home directory on the same machine. I rebuilt the two after cleaning. Then, I ran `mpirun -np {1,2,4} ./laghost --input ./defaults_SL.cfg` only to see the PCG solver fail to converge. 

Please let me know if I'm supposed to use different mfem and Laghost in your home directory.

**Comment by echoi** [comment]
@sungho91 [The commit 9d4d79 made on Nov 13 2023](https://github.com/sungho91/Laghost/commit/9d4d79b32901dd2faace5b160965362cbf66b517) works with the latest mfem meaning it successfully runs through with the default input.

I verified that the [next commit 2ec7c3](https://github.com/sungho91/Laghost/commit/2ec7c38ba651522a3161708d28c46f94ec85db9e) made on Jan. 11, 2024 also works: The default run <s>hangs after this screen output</s> was seen to progress in time.

The [next commit  9b6726](https://github.com/sungho91/Laghost/commit/9b6726c04d2815c88f174bc3efe2ffb3d2d6f89b) also progresses in time with the latest mfem and default.cfg. MFEM's `vector.[cpp,hpp]` had to be modified, though.

The [next commit 1886b4](https://github.com/sungho91/Laghost/commit/1886b4c538aa46c130f45ac87c581e3868232b4c) fails to run with the latest mfem and default.cfg:
```SHELL
simulation starts


Verification failed: (IsFinite(nom)) is false:
 --> nom = nan
 ... in function: virtual void mfem::CGSolver::Mult(const mfem::Vector&, mfem::Vector&) const
 ... in file: linalg/solvers.cpp:747

--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD
with errorcode 1.

NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
```
The failing `CGSolver::Mult` is called in `void LagrangianGeoOperator::SolveVelocity(const Vector &S, Vector &dS_dt) const`: Line 670, `cg.Mult(B, X);`.

@sungho91 The more recent commits are not pervasive and apparently unrelated to the failing CGSolver. We will need to focus our debugging effort on the [commit 1886b4](https://github.com/sungho91/Laghost/commit/1886b4c538aa46c130f45ac87c581e3868232b4c).

**Comment by echoi** [comment]
I just pushed fixes (e5863ff) for the bugs introduced in the commit 1886b4c. 

One question for @sungho91 : `mscale` was set to 1e16 in `defaults.cfg`. The value somehow led to negative `min_detJ` error. So, I pushed a much smaller value, 1e6. The new value let the default model run serially and in parallel. However, **is this acceptable value?** In the light of the mass scaling factors used in geoflac and DynEarthSol, 1e16 seems extreme.

**Comment by sungho91** [comment]
@echoi, I couldn't follow this issue because I can't remember details in the old commit. The first thing I can say is that the 1e16 mass scaling method is the old way I used to do it. However, one thing I'm sure of is that this value is not extreme. Mass scaling is somehow expressed in powers of 2 and multiplied with velocity or modulus in FLAC, so the resultant mass is approximately ~1e-16, as far as I remember. So, if you want to use same mass scaling way used in FLAC, you should use the latest or maybe later than April or May, 2024.

**Comment by echoi** [comment]
This issue has been fixed by the commit e5863ff5.
